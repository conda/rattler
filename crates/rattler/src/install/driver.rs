use std::{
    borrow::Borrow,
    collections::{HashMap, HashSet},
    path::{Path, PathBuf},
    sync::{Arc, Mutex, MutexGuard},
};

use indexmap::IndexSet;
use itertools::Itertools;
use rattler_conda_types::{prefix_record::PathType, PackageRecord, PrefixRecord};
use simple_spawn_blocking::{tokio::run_blocking_task, Cancelled};
use thiserror::Error;
use tokio::sync::{AcquireError, OwnedSemaphorePermit, Semaphore};

use super::{
    clobber_registry::{ClobberError, ClobberRegistry, ClobberedPath},
    link_script::{PrePostLinkError, PrePostLinkResult},
    unlink::{recursively_remove_empty_directories, UnlinkError},
    Transaction,
};
use crate::install::link_script::LinkScriptError;

/// Packages can mostly be installed in isolation and therefore in parallel.
/// However, when installing a large number of packages at the same time the
/// different installation tasks start competing for resources. The
/// [`InstallDriver`] helps to assist in making sure that tasks don't starve
/// each other from resource as well as making sure that due to the large number
/// of requests the process doesn't try to acquire more resources than the
/// system has available.
pub struct InstallDriver {
    io_concurrency_semaphore: Option<Arc<Semaphore>>,
    clobber_registry: Arc<Mutex<ClobberRegistry>>,
    execute_link_scripts: bool,
}

impl Default for InstallDriver {
    fn default() -> Self {
        Self::builder()
            .execute_link_scripts(false)
            .with_io_concurrency_limit(100)
            .finish()
    }
}

/// A builder to configure a new `InstallDriver`.
#[derive(Debug, Default)]
pub struct InstallDriverBuilder {
    io_concurrency_semaphore: Option<Arc<Semaphore>>,
    clobber_registry: Option<ClobberRegistry>,
    execute_link_scripts: bool,
}

/// The result of the post-processing step.
#[derive(Debug)]
pub struct PostProcessResult {
    /// The result of running the post link scripts. This is only present if
    /// running the scripts is allowed.
    pub post_link_result: Option<Result<PrePostLinkResult, LinkScriptError>>,

    /// The paths that were clobbered during the installation process.
    pub clobbered_paths: HashMap<PathBuf, ClobberedPath>,
}

/// An error that might have occurred during post-processing
#[derive(Debug, Error)]
pub enum PostProcessingError {
    #[error("failed to unclobber clobbered files")]
    ClobberError(#[from] ClobberError),

    /// Failed to determine the currently installed packages.
    #[error("failed to determine the installed packages")]
    FailedToDetectInstalledPackages(#[source] std::io::Error),
}

impl InstallDriverBuilder {
    /// Sets an optional IO concurrency limit. This is used to make sure
    /// that the system doesn't acquire more IO resources than the system has
    /// available.
    pub fn with_io_concurrency_limit(self, limit: usize) -> Self {
        Self {
            io_concurrency_semaphore: Some(Arc::new(Semaphore::new(limit))),
            ..self
        }
    }

    /// Sets an optional IO concurrency semaphore. This is used to make sure
    /// that the system doesn't acquire more IO resources than the system has
    /// available.
    pub fn with_io_concurrency_semaphore(self, io_concurrency_semaphore: Arc<Semaphore>) -> Self {
        Self {
            io_concurrency_semaphore: Some(io_concurrency_semaphore),
            ..self
        }
    }

    /// Sets the prefix records that are present in the current environment.
    /// This is used to initialize the clobber registry.
    pub fn with_prefix_records<'i>(
        self,
        prefix_records: impl IntoIterator<Item = &'i PrefixRecord>,
    ) -> Self {
        Self {
            clobber_registry: Some(ClobberRegistry::new(prefix_records)),
            ..self
        }
    }

    /// Sets whether to execute link scripts or not.
    pub fn execute_link_scripts(self, execute_link_scripts: bool) -> Self {
        Self {
            execute_link_scripts,
            ..self
        }
    }

    pub fn finish(self) -> InstallDriver {
        InstallDriver {
            io_concurrency_semaphore: self.io_concurrency_semaphore,
            clobber_registry: self
                .clobber_registry
                .map(Mutex::new)
                .map(Arc::new)
                .unwrap_or_default(),
            execute_link_scripts: self.execute_link_scripts,
        }
    }
}

impl InstallDriver {
    /// Constructs a builder to configure a new `InstallDriver`.
    pub fn builder() -> InstallDriverBuilder {
        InstallDriverBuilder::default()
    }

    /// Returns a permit that will allow the caller to perform IO operations.
    /// This is used to make sure that the system doesn't try to acquire
    /// more IO resources than the system has available.
    pub async fn acquire_io_permit(&self) -> Result<Option<OwnedSemaphorePermit>, AcquireError> {
        match self.io_concurrency_semaphore.clone() {
            None => Ok(None),
            Some(semaphore) => Ok(Some(semaphore.acquire_owned().await?)),
        }
    }

    /// Return a locked reference to the paths registry. This is used to make
    /// sure that the same path is not installed twice.
    pub fn clobber_registry(&self) -> MutexGuard<'_, ClobberRegistry> {
        self.clobber_registry.lock().unwrap()
    }

    /// Call this before any packages are installed to perform any pre
    /// processing that is required.
    pub fn pre_process<Old: Borrow<PrefixRecord>, New>(
        &self,
        transaction: &Transaction<Old, New>,
        target_prefix: &Path,
    ) -> Result<Option<PrePostLinkResult>, PrePostLinkError> {
        if self.execute_link_scripts {
            match self.run_pre_unlink_scripts(transaction, target_prefix) {
                Ok(res) => {
                    return Ok(Some(res));
                }
                Err(e) => {
                    tracing::error!("Error running pre-unlink scripts: {:?}", e);
                }
            }
        }

        Ok(None)
    }

    /// Runs a blocking task that will execute on a separate thread. The task is
    /// not started until an IO permit is acquired. This is used to make
    /// sure that the system doesn't try to acquire more IO resources than
    /// the system has available.
    pub async fn run_blocking_io_task<
        T: Send + 'static,
        E: Send + From<Cancelled> + 'static,
        F: FnOnce() -> Result<T, E> + Send + 'static,
    >(
        &self,
        f: F,
    ) -> Result<T, E> {
        let permit = self.acquire_io_permit().await.map_err(|_err| Cancelled)?;

        run_blocking_task(move || {
            let _permit = permit;
            f()
        })
        .await
    }

    /// Call this after all packages have been installed to perform any post
    /// processing that is required.
    ///
    /// This function will select a winner among multiple packages that might
    /// write to a single package and will also execute any
    /// `post-link.sh/bat` scripts
    pub fn post_process<Old: Borrow<PrefixRecord> + AsRef<New>, New: AsRef<PackageRecord>>(
        &self,
        transaction: &Transaction<Old, New>,
        target_prefix: &Path,
    ) -> Result<PostProcessResult, PostProcessingError> {
        let prefix_records = PrefixRecord::collect_from_prefix(target_prefix)
            .map_err(PostProcessingError::FailedToDetectInstalledPackages)?;

        let required_packages =
            PackageRecord::sort_topologically(prefix_records.iter().collect::<Vec<_>>());

        self.remove_empty_directories(transaction, &prefix_records, target_prefix)
            .unwrap_or_else(|e| {
                tracing::warn!("Failed to remove empty directories: {} (ignored)", e);
            });

        let clobbered_paths = self
            .clobber_registry()
            .unclobber(&required_packages, target_prefix)?;

        let post_link_result = if self.execute_link_scripts {
            Some(self.run_post_link_scripts(transaction, &required_packages, target_prefix))
        } else {
            None
        };

        Ok(PostProcessResult {
            post_link_result,
            clobbered_paths,
        })
    }

    /// Remove all empty directories that are not part of the new prefix
    /// records.
    pub fn remove_empty_directories<Old: Borrow<PrefixRecord>, New>(
        &self,
        transaction: &Transaction<Old, New>,
        new_prefix_records: &[PrefixRecord],
        target_prefix: &Path,
    ) -> Result<(), UnlinkError> {
        let mut keep_directories = HashSet::new();

        // find all forced directories in the prefix records
        for record in new_prefix_records {
            for paths in record.paths_data.paths.iter() {
                if paths.path_type == PathType::Directory {
                    let path = target_prefix.join(&paths.relative_path);
                    keep_directories.insert(path);
                }
            }
        }

        // find all removed directories
        for record in transaction.removed_packages().map(Borrow::borrow) {
            let mut removed_directories = HashSet::new();

            for paths in record.paths_data.paths.iter() {
                if paths.path_type != PathType::Directory {
                    if let Some(parent) = paths.relative_path.parent() {
                        removed_directories.insert(parent);
                    }
                }
            }

            let is_python_noarch = record.repodata_record.package_record.noarch.is_python();

            // Sort the directories by length, so that we delete the deepest directories
            // first.
            let mut directories: IndexSet<_> = removed_directories.into_iter().sorted().collect();

            while let Some(directory) = directories.pop() {
                let directory_path = target_prefix.join(directory);
                let removed_until = recursively_remove_empty_directories(
                    &directory_path,
                    target_prefix,
                    is_python_noarch,
                    &keep_directories,
                )?;

                // The directory is not empty which means our parent directory is also not
                // empty, recursively remove the parent directory from the set
                // as well.
                while let Some(parent) = removed_until.parent() {
                    if !directories.shift_remove(parent) {
                        break;
                    }
                }
            }
        }

        Ok(())
    }
}
