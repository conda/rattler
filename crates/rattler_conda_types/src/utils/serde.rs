use chrono::{DateTime, Utc};
use fxhash::FxHashMap;
use serde::de::Error as _;
use serde::ser::Error;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_with::{DeserializeAs, SerializeAs};
use std::collections::BTreeMap;
use std::marker::PhantomData;
use std::path::{Path, PathBuf};
use url::Url;

/// A helper struct that serializes Paths in a normalized way.
/// - Backslashes are replaced with forward-slashes.
pub(crate) struct NormalizedPath;

impl<P: AsRef<Path>> SerializeAs<P> for NormalizedPath {
    fn serialize_as<S>(source: &P, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        match source.as_ref().to_str() {
            Some(s) => s.replace('\\', "/").serialize(serializer),
            None => Err(S::Error::custom("path contains invalid UTF-8 characters")),
        }
    }
}

impl<'de> DeserializeAs<'de, PathBuf> for NormalizedPath {
    fn deserialize_as<D>(deserializer: D) -> Result<PathBuf, D::Error>
    where
        D: Deserializer<'de>,
    {
        PathBuf::deserialize(deserializer)
    }
}

/// Deserialize a sequence into `Vec<T>` but filter `None` values.
pub(crate) struct VecSkipNone<T>(PhantomData<T>);

impl<'de, T, I> DeserializeAs<'de, Vec<T>> for VecSkipNone<I>
where
    I: DeserializeAs<'de, Vec<Option<T>>>,
{
    fn deserialize_as<D>(deserializer: D) -> Result<Vec<T>, D::Error>
    where
        D: Deserializer<'de>,
    {
        Ok(I::deserialize_as(deserializer)?
            .into_iter()
            .flatten()
            .collect())
    }
}

/// A helper type parser that tries to parse Urls that could be malformed.
pub(crate) struct LossyUrl;

impl<'de> DeserializeAs<'de, Option<Url>> for LossyUrl {
    fn deserialize_as<D>(deserializer: D) -> Result<Option<Url>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let str = match Option::<String>::deserialize(deserializer)? {
            Some(url) => url,
            None => return Ok(None),
        };
        let url = match Url::parse(&str) {
            Ok(url) => url,
            Err(e) => {
                tracing::warn!("unable to parse '{}' as an URL: {e}. Skipping...", str);
                return Ok(None);
            }
        };
        Ok(Some(url))
    }
}

/// A helper type that parses a string either as a string or a vector of strings.
pub(crate) struct MultiLineString;

impl<'de> DeserializeAs<'de, String> for MultiLineString {
    fn deserialize_as<D>(deserializer: D) -> Result<String, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        #[serde(untagged)]
        enum Inner {
            String(String),
            Multi(Vec<String>),
        }

        Ok(match Inner::deserialize(deserializer)? {
            Inner::String(s) => s,
            Inner::Multi(s) => s.join("\n"),
        })
    }
}

pub(crate) struct Timestamp;

impl<'de> DeserializeAs<'de, chrono::DateTime<chrono::Utc>> for Timestamp {
    fn deserialize_as<D>(deserializer: D) -> Result<chrono::DateTime<chrono::Utc>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let timestamp = i64::deserialize(deserializer)?;

        // Convert from milliseconds to seconds
        let microseconds = if timestamp > 253_402_300_799 {
            timestamp * 1_000
        } else {
            timestamp * 1_000_000
        };

        // Convert the timestamp to a UTC timestamp
        chrono::DateTime::from_timestamp_micros(microseconds)
            .ok_or_else(|| D::Error::custom("got invalid timestamp, timestamp out of range"))
    }
}

impl SerializeAs<chrono::DateTime<chrono::Utc>> for Timestamp {
    fn serialize_as<S>(source: &DateTime<Utc>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        // Convert the date to a timestamp
        let timestamp: i64 = source.timestamp_millis();

        // Determine the precision of the timestamp.
        let timestamp = if timestamp % 1000 == 0 {
            timestamp / 1000
        } else {
            timestamp
        };

        // Serialize the timestamp
        timestamp.serialize(serializer)
    }
}

/// A helper struct to deserialize types from a string without checking the string.
pub struct DeserializeFromStrUnchecked;

/// A helper function used to sort map alphabetically when serializing.
pub(crate) fn sort_map_alphabetically<T: Serialize, S: serde::Serializer>(
    value: &FxHashMap<String, T>,
    serializer: S,
) -> Result<S::Ok, S::Error> {
    value
        .iter()
        .collect::<BTreeMap<_, _>>()
        .serialize(serializer)
}
