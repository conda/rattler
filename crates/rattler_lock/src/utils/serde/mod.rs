mod match_spec_map_or_vec;
mod ordered;
mod pep440_map_or_vec;
mod timestamp;
pub(crate) mod url_or_path;
pub(crate) use match_spec_map_or_vec::MatchSpecMapOrVec;
pub(crate) use ordered::Ordered;
pub(crate) use pep440_map_or_vec::Pep440MapOrVec;
pub(crate) use timestamp::Timestamp;
