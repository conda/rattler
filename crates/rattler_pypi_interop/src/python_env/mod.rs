//! Module for working with python environments.
//!
//! Contains functionality for querying python environments, including collecting metadata
//! about packages installed in a python environment.
//!
//! Example of querying a python environment for installed packages (i.e. distributions):
//!
//! ```rust
//! use std::path::Path;
//! use rattler_pypi_interop::types::InstallPaths;
//! use rattler_pypi_interop::python_env::{
//!     find_distributions_in_venv,
//!     Distribution,
//! };
//!
//! fn main() -> Result<(), Box<dyn std::error::Error>> {
//!     let venv_path = Path::new(env!("CARGO_MANIFEST_DIR"))
//!         .join(Path::new("test-data/find_distributions"));
//!     let windows = false;
//!     let install_paths = InstallPaths::for_venv((3, 8, 5), windows);
//!
//!     let distributions = find_distributions_in_venv(&venv_path, &install_paths)?;
//!
//!     /// Print all distributions found in the virtual environment.
//!     println!("{:?}", distributions);
//!
//!     Ok(())
//! }
//! ```

mod byte_code_compiler;
mod distribution_finder;
mod env_markers;
mod system_python;
mod tags;
mod uninstall;
mod venv;

pub use tags::{WheelTag, WheelTags};

pub use byte_code_compiler::{ByteCodeCompiler, CompilationError, SpawnCompilerError};
pub use distribution_finder::{
    find_distributions_in_directory, find_distributions_in_venv, Distribution,
    FindDistributionError,
};
pub use env_markers::Pep508EnvMakers;
pub(crate) use system_python::{system_python_executable, FindPythonError};
pub use system_python::{ParsePythonInterpreterVersionError, PythonInterpreterVersion};
pub use uninstall::{uninstall_distribution, UninstallDistributionError};
pub use venv::PythonLocation;
