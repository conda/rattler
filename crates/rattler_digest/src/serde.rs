//! Provides custom serialization/deserialization functions for [`Output`] of a [`Digest`]
//! Use the struct [`SerializableHash`] to easily serialize the digest.
//!
//! # Example:
//!
//! ```
//!
//! use rattler_digest::serde::SerializableHash;
//!
//! let hash = SerializableHash::<sha2::Sha256>(
//! rattler_digest::parse_digest_from_hex::<sha2::Sha256>("fe51de6107f9edc7aa4f786a70f4a883943bc9d39b3bb7307c04c41410990726").unwrap());
//! let str = serde_json::to_string(&hash).unwrap();
//! let hash: SerializableHash<sha2::Sha256> = serde_json::from_str(&str).unwrap();
//!
//! ```
//!
use digest::{Digest, Output};
use serde::de::Error;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_with::{DeserializeAs, SerializeAs};
use std::borrow::Cow;
use std::fmt::LowerHex;
use std::ops::Deref;

/// Deserialize the [`Output`] of a [`Digest`].
///
/// If the deserializer is human-readable, it will parse the digest from a hex
/// string. Otherwise, it will deserialize raw bytes.
pub fn deserialize<'de, D, Dig: Digest>(deserializer: D) -> Result<Output<Dig>, D::Error>
where
    D: Deserializer<'de>,
{
    if deserializer.is_human_readable() {
        let str = Cow::<'de, str>::deserialize(deserializer)?;
        super::parse_digest_from_hex::<Dig>(str.as_ref())
            .ok_or_else(|| Error::custom("failed to parse digest"))
    } else {
        Output::<Dig>::deserialize(deserializer)
    }
}

/// Serializes the [`Output`] of a [`Digest`].
///
/// If the serializer is human-readable, it will write the digest as a hex
/// string. Otherwise, it will deserialize raw bytes.
pub fn serialize<'a, S: Serializer, Dig: Digest>(
    digest: &'a Output<Dig>,
    s: S,
) -> Result<S::Ok, S::Error>
where
    &'a Output<Dig>: LowerHex,
{
    if s.is_human_readable() {
        format!("{digest:x}").serialize(s)
    } else {
        digest.serialize(s)
    }
}

/// Wrapper type for easily serializing a Hash
#[derive(Debug, Clone)]
pub struct SerializableHash<T: Digest>(pub Output<T>);

impl<T: Digest> Serialize for SerializableHash<T>
where
    Output<T>: LowerHex,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize::<S, T>(&self.0, serializer)
    }
}

impl<'de, T: Digest + Default> Deserialize<'de> for SerializableHash<T> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let hash_output: Output<T> = deserialize::<D, T>(deserializer)?;
        Ok(SerializableHash(hash_output))
    }
}

impl<T: Digest> From<Output<T>> for SerializableHash<T> {
    fn from(output: Output<T>) -> Self {
        SerializableHash(output)
    }
}

impl<T: Digest> From<SerializableHash<T>> for Output<T> {
    fn from(s: SerializableHash<T>) -> Self {
        s.0
    }
}

impl<T: Digest> Deref for SerializableHash<T> {
    type Target = Output<T>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T: Digest> SerializeAs<Output<T>> for SerializableHash<T>
where
    for<'a> &'a Output<T>: LowerHex,
{
    fn serialize_as<S>(source: &Output<T>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize::<S, T>(source, serializer)
    }
}

impl<'de, T: Digest + Default> DeserializeAs<'de, Output<T>> for SerializableHash<T> {
    fn deserialize_as<D>(deserializer: D) -> Result<Output<T>, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserialize::<D, T>(deserializer)
    }
}

#[cfg(test)]
mod test {
    use crate::serde::SerializableHash;

    #[test]
    pub fn test_serializable_hash() {
        let hash = SerializableHash::<sha2::Sha256>(
            crate::parse_digest_from_hex::<sha2::Sha256>(
                "fe51de6107f9edc7aa4f786a70f4a883943bc9d39b3bb7307c04c41410990726",
            )
            .unwrap(),
        );
        let str = serde_json::to_string(&hash).unwrap();
        let _hash: SerializableHash<sha2::Sha256> = serde_json::from_str(&str).unwrap();
    }
}
