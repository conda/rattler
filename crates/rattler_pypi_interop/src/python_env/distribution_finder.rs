//! This module implements logic to locate so called python distributions (installed artifacts)
//! in an environment.
//!
//! The implementation is based on the <https://packaging.python.org/en/latest/specifications/recording-installed-packages>
//! which is based on [PEP 376](https://peps.python.org/pep-0376/) and [PEP 627](https://peps.python.org/pep-0627/).

use crate::install::InstallPaths;
use crate::python_env::WheelTag;
use crate::{types::NormalizedPackageName, types::PackageName, types::RFC822ish};
use fs_err as fs;
use indexmap::IndexSet;
use itertools::Itertools;
use pep440_rs::Version;
use serde::{Deserialize, Serialize};
use std::{
    ffi::OsStr,
    path::{Path, PathBuf},
    str::FromStr,
};
use thiserror::Error;

/// Information about a distribution found by `find_distributions_in_venv`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Distribution {
    /// The name of the distribution
    pub name: NormalizedPackageName,

    /// The version of the distribution
    pub version: Version,

    /// The installer that was responsible for installing the distribution
    pub installer: Option<String>,

    /// The path to the .dist-info directory relative to the root of the environment.
    pub dist_info: PathBuf,

    /// The specific tags of the distribution that was installed or `None` if this information
    /// could not be retrieved.
    pub tags: Option<IndexSet<WheelTag>>,
}

/// An error that can occur when running `find_distributions_in_venv`.
#[derive(Debug, Error)]
pub enum FindDistributionError {
    /// An IO error occurred
    #[error(transparent)]
    IoError(#[from] std::io::Error),

    /// Failed to parse a WHEEL file
    #[error("failed to parse '{0}'")]
    FailedToParseWheel(PathBuf, #[source] <RFC822ish as FromStr>::Err),

    /// Failed to parse WHEEL tags
    #[error("failed to parse wheel tag {0}")]
    FailedToParseWheelTag(String),
}

/// Locates the python distributions (packages) that have been installed in the specified directory.
///
/// When packages are installed in a venv they are installed in specific directories. Use the
/// [`find_distributions_in_venv`] if you don't want to deal with determining the proper directory.
///
/// Any path in the results is relative to `search_dir`.
pub fn find_distributions_in_directory(
    search_dir: &Path,
) -> Result<Vec<Distribution>, FindDistributionError> {
    let mut result = Vec::new();
    for entry in search_dir.read_dir()? {
        let entry = entry?;
        if entry.file_type()?.is_dir() {
            if let Some(dist) = analyze_distribution(entry.path())? {
                result.push(Distribution {
                    dist_info: pathdiff::diff_paths(&dist.dist_info, search_dir)
                        .unwrap_or(dist.dist_info),
                    ..dist
                })
            }
        }
    }

    Ok(result)
}

/// Locates the python distributions (packages) that have been installed in the virtualenv rooted at
/// `root`.
pub fn find_distributions_in_venv(
    root: &Path,
    paths: &InstallPaths,
) -> Result<Vec<Distribution>, FindDistributionError> {
    // We will look for distributions in the purelib/platlib directories
    let locations = [paths.purelib(), paths.platlib()]
        .into_iter()
        .map(|p| root.join(p))
        .unique()
        .filter(|p| p.is_dir());

    let mut results = Vec::new();
    for dir in locations {
        // Find distributions in the directory.
        let distributions = find_distributions_in_directory(&dir)?;

        // Modify the paths in the result to be relative to the root of the environment instead of
        // to the search directory.
        let dir_relative_path = pathdiff::diff_paths(&dir, root).unwrap_or_else(|| dir.clone());
        results.extend(distributions.into_iter().map(|dist| Distribution {
            // Make the paths relative to the root
            dist_info: dir_relative_path.join(dist.dist_info),
            ..dist
        }));
    }

    Ok(results)
}

/// Analyzes a `.dist-info` directory to see if it actually contains a python distribution (package).
fn analyze_distribution(
    dist_info_path: PathBuf,
) -> Result<Option<Distribution>, FindDistributionError> {
    let Some((name, version)) = dist_info_path
        .file_name()
        .and_then(OsStr::to_str)
        .and_then(|n| n.strip_suffix(".dist-info"))
        .and_then(|n| n.split_once('-'))
    else {
        // If we are unable to parse the distribution name we just skip.
        return Ok(None);
    };

    // Check if the METADATA file is present. This is the only file that is mandatory, so if its
    // missing this folder is not a python distribution.
    if !dist_info_path.join("METADATA").is_file() {
        return Ok(None);
    }

    // Parse the name
    let Ok(name) = PackageName::from_str(name) else {
        // If the package name cannot be parsed, just skip
        return Ok(None);
    };

    // Parse the version
    let Ok(version) = Version::from_str(version) else {
        // If the version cannot be parsed, just skip
        return Ok(None);
    };

    // Try to read the INSTALLER file from the distribution directory
    let installer = fs::read_to_string(dist_info_path.join("INSTALLER"))
        .map(|i| i.trim().to_owned())
        .ok();

    // Check if there is a WHEEL file from where we can read tags
    let wheel_path = dist_info_path.join("WHEEL");
    let tags = if wheel_path.is_file() {
        let mut parsed = RFC822ish::from_str(&fs::read_to_string(&wheel_path)?)
            .map_err(move |e| FindDistributionError::FailedToParseWheel(wheel_path, e))?;

        Some(
            parsed
                .take_all("Tag")
                .into_iter()
                .map(|tag| {
                    WheelTag::from_compound_string(&tag)
                        .map_err(|_| FindDistributionError::FailedToParseWheelTag(tag))
                })
                .flatten_ok()
                .collect::<Result<IndexSet<_>, _>>()?,
        )
    } else {
        None
    };

    Ok(Some(Distribution {
        dist_info: dist_info_path,
        name: name.into(),
        version,
        installer,
        tags,
    }))
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_find_distributions() {
        // Describe the virtual environment
        let venv_path =
            Path::new(env!("CARGO_MANIFEST_DIR")).join("../../test-data/find_distributions/");
        let install_paths = InstallPaths::for_venv((3, 8, 5), true);

        // Find all distributions
        let mut distributions = find_distributions_in_venv(&venv_path, &install_paths).unwrap();

        // Sort to get consistent ordering across platforms
        distributions.sort_by(|a, b| a.name.cmp(&b.name));

        insta::assert_ron_snapshot!(distributions, {
            "[].dist_info" => insta::dynamic_redaction(move |value, _path| {
                value.as_str().unwrap().replace('\\', "/")
            }),
        });
    }
}
