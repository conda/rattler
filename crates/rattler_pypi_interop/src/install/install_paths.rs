use crate::python_env::PythonInterpreterVersion;
use std::borrow::Cow;
use std::path::{Path, PathBuf};

/// A struct of installation categories to where they should be stored relative to the
/// installation destination.
#[derive(Debug, Clone)]
pub struct InstallPaths {
    purelib: PathBuf,
    platlib: PathBuf,
    scripts: PathBuf,
    data: PathBuf,
    headers: PathBuf,
    windows: bool,
}

impl InstallPaths {
    /// Populates mappings of installation targets for a virtualenv layout. The mapping depends on
    /// the python version and whether or not the installation targets windows. Specifically on
    /// windows some of the paths are different. :shrug:
    pub fn for_venv<V: Into<PythonInterpreterVersion>>(version: V, windows: bool) -> Self {
        let version = version.into();

        let site_packages = if windows {
            Path::new("Lib").join("site-packages")
        } else {
            Path::new("lib").join(format!(
                "python{}.{}/site-packages",
                version.major, version.minor
            ))
        };
        let scripts = if windows {
            PathBuf::from("Scripts")
        } else {
            PathBuf::from("bin")
        };

        // Data should just be the root of the venv
        let data = PathBuf::from("");

        // purelib and platlib locations are not relevant when using venvs
        // https://stackoverflow.com/a/27882460/3549270
        Self {
            purelib: site_packages.clone(),
            platlib: site_packages,
            scripts,
            data,
            windows,
            headers: PathBuf::from("include"),
        }
    }

    /// Determines whether this is a windows InstallPath
    pub fn is_windows(&self) -> bool {
        self.windows
    }

    /// Returns the site-packages location. This is done by searching for the purelib location.
    pub fn site_packages(&self) -> &Path {
        &self.purelib
    }

    /// Reference to pure python library location.
    pub fn purelib(&self) -> &Path {
        &self.purelib
    }

    /// Reference to platform specific library location.
    pub fn platlib(&self) -> &Path {
        &self.platlib
    }

    /// Returns the binaries location.
    pub fn scripts(&self) -> &Path {
        &self.scripts
    }

    /// Returns the location of the data directory
    pub fn data(&self) -> &Path {
        &self.data
    }

    /// Returns the location of the include directory
    pub fn include(&self) -> PathBuf {
        if self.windows {
            PathBuf::from("Include")
        } else {
            PathBuf::from("include")
        }
    }

    /// Returns the location of the headers directory. The location of headers is specific to a
    /// distribution name.
    pub fn headers(&self, distribution_name: &str) -> PathBuf {
        self.headers.join(distribution_name)
    }

    /// Matches the different categories to their install paths.
    pub fn match_category(&self, category: &str, distribution_name: &str) -> Option<Cow<Path>> {
        match category {
            "purelib" => Some(self.purelib().into()),
            "platlib" => Some(self.platlib().into()),
            "scripts" => Some(self.scripts().into()),
            "data" => Some(self.data().into()),
            "headers" => Some(self.headers(distribution_name).into()),
            &_ => None,
        }
    }
}
