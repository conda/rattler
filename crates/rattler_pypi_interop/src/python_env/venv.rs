//! Module that helps with allowing in the creation of python virtual environments.
//! Now just use the python venv command to create the virtual environment.
//! Later on we can look into actually creating the environment by linking to the python library,
//! and creating the necessary files. See: [VEnv](https://packaging.python.org/en/latest/specifications/virtual-environments/#declaring-installation-environments-as-python-virtual-environments)
use crate::python_env::{
    system_python_executable, FindPythonError, ParsePythonInterpreterVersionError,
    PythonInterpreterVersion,
};
use std::fmt::Debug;
use std::path::PathBuf;

/// Specifies where to find the python executable
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub enum PythonLocation {
    /// Use system interpreter
    #[default]
    System,
    /// Use custom interpreter
    Custom(PathBuf),

    /// Use custom interpreter with version
    CustomWithVersion(PathBuf, PythonInterpreterVersion),
}

impl PythonLocation {
    /// Location of python executable
    pub fn executable(&self) -> Result<PathBuf, FindPythonError> {
        match self {
            PythonLocation::System => system_python_executable().cloned(),
            PythonLocation::Custom(path) | PythonLocation::CustomWithVersion(path, _) => {
                Ok(path.clone())
            }
        }
    }

    /// Get python version from executable
    pub fn version(&self) -> Result<PythonInterpreterVersion, ParsePythonInterpreterVersionError> {
        match self {
            PythonLocation::System => PythonInterpreterVersion::from_system(),
            PythonLocation::CustomWithVersion(_, version) => Ok(version.clone()),
            PythonLocation::Custom(path) => PythonInterpreterVersion::from_path(path),
        }
    }
}
