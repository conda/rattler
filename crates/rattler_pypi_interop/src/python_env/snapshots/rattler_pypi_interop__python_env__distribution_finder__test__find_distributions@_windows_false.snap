---
source: crates/rattler_pypi_interop/src/python_env/distribution_finder.rs
expression: distributions
---
[
  Distribution(
    name: "click",
    version: "7.1.2",
    installer: Some("pip"),
    dist_info: "lib/python3.8/site-packages/click-7.1.2.dist-info",
    tags: Some([
      "py2-none-any",
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "flask",
    version: "1.1.4",
    installer: Some("pip"),
    dist_info: "lib/python3.8/site-packages/Flask-1.1.4.dist-info",
    tags: Some([
      "py2-none-any",
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "itsdangerous",
    version: "1.1.0",
    installer: Some("pip"),
    dist_info: "lib/python3.8/site-packages/itsdangerous-1.1.0.dist-info",
    tags: Some([
      "py2-none-any",
      "py3-none-any",
    ]),
  ),
]
